### PR TITLE
Improved adding relation on OSSMailScanner

### PR DESCRIPTION
**Pull request by adudekyeti**
Improved adding relation on OSSMailScanner

### CODE REVIEW

File: modules/OSSMailScanner/models/PrefixScannerAction.php [hunk 1]
```diff
@@ -20,37 +20,44 @@ abstract class OSSMailScanner_PrefixScannerAction_Model
...
+			}
+			$dataReader->close();
+			if (!empty($returnIds)) {
+				$recordId = $returnIds;
+			} else {
+				$this->prefix = \App\Fields\Email::findRecordNumber($this->mail->get('subject'), $this->moduleName);
+				if ($this->prefix) {
+					$recordId = $this->add();
+				} elseif (\App\Config::module('OSSMailScanner', 'SEARCH_PREFIX_IN_BODY') && $this->prefix = \App\Fields\Email::findRecordNumber($this->mail->get('body'), $this->moduleName, true)) {
+					if ($this->prefix) {
```
Comment (rskrzypczak) [inline modules/OSSMailScanner/models/PrefixScannerAction.php]: remove this condition (duplicate)

File: modules/OSSMailScanner/models/PrefixScannerAction.php [hunk 2]
```diff
@@ -20,37 +20,44 @@ abstract class OSSMailScanner_PrefixScannerAction_Model
...
-		while ($crmId = $dataReader->readColumn(0)) {
-			if (\App\Record::getType($crmId) === $this->moduleName) {
-				$returnIds[] = $crmId;
+		$recordId = 0;
+		if ($mailId) {
+			$returnIds = [];
+			$query = (new \App\Db\Query())->select(['crmid'])->from('vtiger_ossmailview_relation')->where(['ossmailviewid' => $mailId]);
+			$dataReader = $query->createCommand()->query();
+			while ($crmId = $dataReader->readColumn(0)) {
+				if (\App\Record::getType($crmId) === $this->moduleName) {
```
Comment (rskrzypczak) [inline modules/OSSMailScanner/models/PrefixScannerAction.php]: remove loop
add join to query and add condition [setype =>$this->moduleName, ['<>',deleted,1]]

File: modules/OSSMailScanner/models/PrefixScannerAction.php [hunk 3]
```diff
@@ -85,4 +92,47 @@ protected function add()
...
+	{
+		$returnIds = [];
+		$crmId = false;
+		$prefixes = implode(',', $this->prefix);
+		if (\App\Cache::has('getRecordByPrefix', $prefixes)) {
+			$crmId = \App\Cache::get('getRecordByPrefix', $prefixes);
+		} else {
+			$crmId = $this->getNewestRecord();
+			if ($crmId) {
+				\App\Cache::save('getRecordByPrefix', $prefixes, $crmId, \App\Cache::LONG);
```
Comment (rskrzypczak) [inline modules/OSSMailScanner/models/PrefixScannerAction.php]: where do you clear this variable?

File: modules/OSSMailScanner/models/PrefixScannerAction.php [hunk 4]
```diff
@@ -85,4 +92,47 @@ protected function add()
...
+	 * Get newest record id of given record prefixes.
+	 *
+	 * @return bool|int
+	 */
+	public function getNewestRecord()
+	{
+		$queryGenerator = new App\QueryGenerator($this->moduleName);
+		$statusFieldName = \App\RecordStatus::getFieldName($this->moduleName);
+		$queryGenerator->addCondition($statusFieldName,
+		\App\RecordStatus::getStates($this->moduleName, \App\RecordStatus::RECORD_STATE_OPEN), 'e', false);
```
Comment (rskrzypczak) [inline modules/OSSMailScanner/models/PrefixScannerAction.php]: What happens if the module does not support the status mechanism?

File: modules/OSSMailScanner/scanneractions/CreatedHelpDesk.php [hunk 5]
```diff
@@ -43,18 +48,18 @@ public function process(OSSMail_Mail_Model $mail)
...
-	public function add(OSSMail_Mail_Model $mail)
+	public function add()
 	{
-		$contactId = (int) $mail->findEmailAdress('fromaddress', 'Contacts', false);
-		$parentId = (int) $mail->findEmailAdress('fromaddress', 'Accounts', false);
+		$contactId = (int) $this->mail->findEmailAdress('fromaddress', 'Contacts', false);
+		$parentId = (int) $this->mail->findEmailAdress('fromaddress', 'Accounts', false);
 		$record = Vtiger_Record_Model::getCleanInstance('HelpDesk');
-
+		if(!$contactId && !$parentId && \App\Process::$requestMode !== 'WebUI' && !\App\Config::module('OSSMailScanner', 'CREATE_TICKET_WITHOUT_CONTACT')){
```
Comment (rskrzypczak) [inline modules/OSSMailScanner/scanneractions/CreatedHelpDesk.php]: > \App\Process::$requestMode !== 'WebUI'

?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: modules/OSSMailScanner/ConfigTemplate.php [hunk 6]
```diff
@@ -11,5 +11,17 @@
 		'description' => 'Add only one mail for multiple recipients, type: boolean',
 		'validation' => '\App\Validator::bool',
 		'sanitization' => '\App\Purifier::bool'
-	]
+	],
+	'SEARCH_PREFIX_IN_BODY' => [
+		'default' => true,
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMailScanner/ConfigTemplate.php:16]:
```suggestion
		'default' => false,
```

File: modules/OSSMailScanner/scanneractions/CreatedHelpDesk.php (hunk 5 above)
Comment (mariuszkrzaczkowski) [reply modules/OSSMailScanner/scanneractions/CreatedHelpDesk.php:60]: ?

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

**Comment by adudekyeti** [comment]
Duplicate #13710

**Pull request closed without merging**